File: 0-1KnapSackProblem/0-1KnapSackProblem.cpp
/*
https://www.geeksforgeeks.org/problems/0-1-knapsack-problem0945/1
*/

//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
class Solution {
  public:
  
    long long dfs(int W, vector<int>&wt, vector<int>&val, int idx, vector<vector<int>>&memo){
        if(idx>=wt.size()) return 0;
        if(memo[W][idx]!=-1) return memo[W][idx];
        long long res=0, pk=0, ntpk=0;
        ntpk=dfs(W, wt, val, idx+1, memo);
        if(wt[idx]<=W){
            pk=val[idx]+dfs(W-wt[idx], wt, val, idx+1, memo);
            res=max(res, max(pk, ntpk));
        }
        res=max(res, max(pk, ntpk));
        return memo[W][idx]=res;
    }
  
    // Function to return max value that can be put in knapsack of capacity W.
    int knapSack(int W, vector<int>& wt, vector<int>& val) {
        int idx=0;
        vector<vector<int>>memo(W+1, vector<int>(wt.size()+1, -1));
        return dfs(W, wt, val, idx, memo);
    }
};



//{ Driver Code Starts.

int main() {
    // taking total testcases
    int t;
    cin >> t;
    cin.ignore();
    while (t--) {
        // reading number of elements and weight
        int n, w;
        vector<int> arr, val, wt, drr;
        string ip;
        int number;
        getline(cin, ip);
        stringstream ss(ip);

        while (ss >> number) {
            arr.push_back(number);
        }

        getline(cin, ip);
        ss.clear();
        ss.str(ip);

        while (ss >> number) {
            val.push_back(number);
        }

        w = arr[0];
        n = val.size();
        getline(cin, ip);
        ss.clear();
        ss.str(ip);

        while (ss >> number) {
            wt.push_back(number);
        }
        Solution ob;
        cout << ob.knapSack(w, wt, val) << endl;
    }
    return 0;
}
// } Driver Code Ends

File: 0-1KnapSackProblem/0-1KnapSackProblem.java
/*
https://www.geeksforgeeks.org/problems/0-1-knapsack-problem0945/1
*/

//{ Driver Code Starts
import java.io.*;
import java.util.*;

class gfg {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine());

        while (t-- > 0) {
            int w = Integer.parseInt(br.readLine());

            String line = br.readLine();
            String[] tokens = line.split(" ");
            List<Integer> array = new ArrayList<>();

            // Parse the tokens into integers and add to the array
            for (String token : tokens) {
                array.add(Integer.parseInt(token));
            }

            int[] val = new int[array.size()];
            int idx = 0;
            for (int i : array) val[idx++] = i;

            String lin = br.readLine();
            String[] toke = lin.split(" ");
            List<Integer> array1 = new ArrayList<>();

            // Parse the tokens into integers and add to the array
            for (String token : toke) {
                array1.add(Integer.parseInt(token));
            }

            int[] wt = new int[array1.size()];
            idx = 0;
            for (int i : array1) wt[idx++] = i;

            // calling method knapSack() of class Solution
            System.out.println(new Solution().knapSack(w, wt, val));
        }
    }
}
// } Driver Code Ends


class Solution {
    
    public static int depthFirstSearch_forArrays(int W, int wt[], int val[], int beginIndex, int[][] memoization){
        if(beginIndex>=wt.length) return 0;
        if(memoization[W][beginIndex]!=-1) return memoization[W][beginIndex];
        int resultantMaxSumValues=0, resultAfterPickingCurrentElement=0, resultAfterNotPickingCurrentElement=0;
        resultAfterNotPickingCurrentElement=depthFirstSearch_forArrays(W, wt, val, beginIndex+1, memoization);
        if(wt[beginIndex]<=W){
            resultAfterPickingCurrentElement=val[beginIndex]+depthFirstSearch_forArrays(W-wt[beginIndex], wt, val, beginIndex+1, memoization);
            resultantMaxSumValues=Math.max(resultantMaxSumValues, Math.max(resultAfterPickingCurrentElement, resultAfterNotPickingCurrentElement));
        }
        resultantMaxSumValues=Math.max(resultantMaxSumValues, Math.max(resultAfterPickingCurrentElement, resultAfterNotPickingCurrentElement));
        return memoization[W][beginIndex]=resultantMaxSumValues;
    }
    
    // Function to return max value that can be put in knapsack of capacity W.
    static int knapSack(int W, int wt[], int val[]) {
        int beginIndex=0;
        int[][] memoization=new int[W+1][wt.length+1];
        for(int i=0; i<=W; ++i){
            for(int j=0; j<=wt.length; ++j){
                memoization[i][j]=-1;
            }
        }
        return depthFirstSearch_forArrays(W, wt, val, beginIndex, memoization);
    }
}
